src/main/scala/com/fulcrumgenomics/bam/ConsensusCaller.scala

```diff
@@ -0,0 +1,369 @@
+    var success = false
+    success |= this.callConsensus(fragments,    readName=fragmentName, readType=Fragment)
+    success |= this.callConsensus(firstOfPair,  readName=pairName,     readType=FirstOfPair)
+    success |= this.callConsensus(secondOfPair, readName=pairName,     readType=SecondOfPair)
+
+    if (success) true
+    else if (this.iter.nonEmpty) this.advance() // try again
+    else false // failed to build a consensus, and no more records in the input iterator
+  }
+
+  /** Adds a SAM record from the underlying iterator to the buffer if either the buffer is empty or the SAM attribute is
+    * the same for the records in the buffer as the next record in the input iterator.  Returns true if a record was
+    * added, false otherwise.
+    */
+  @annotation.tailrec
+  private def getNextGroupOfRecords(buffer: ListBuffer[SAMRecord] = new ListBuffer[SAMRecord], attr: String): ListBuffer[SAMRecord] = {
+    if (this.iter.isEmpty || this.iter.head.toAttr != attr) return buffer
+    val rec = this.iter.next()
+    progress.map(_.record(rec))
+    buffer.append(rec)
+    getNextGroupOfRecords(buffer=buffer, attr=attr)
+  }
+
+  /** Adds a method to the SAMRecord class to retrieve the attribute for the UMI identifier. */
+  implicit private class SAMRecordToString(r: SAMRecord) {
+    def toAttr: String = r.getStringAttribute(options.attribute)
+  }
+
+  /** Split records into those that should make a single-end consensus read, first of pair consensus read,
+    * and second of pair consensus read, respectively.  The default method is to use the SAM flag to find
+    * unpaired reads, first of pair reads, and second of pair reads.
+    */
+  protected def subGroupRecords(records: Seq[SAMRecord]): (Seq[SAMRecord], Seq[SAMRecord],Seq[SAMRecord]) = {
+    val fragments    = records.filter { rec => !rec.getReadPairedFlag }
+    val firstOfPair  = records.filter { rec => rec.getReadPairedFlag && rec.getFirstOfPairFlag }
+    val secondOfPair = records.filter { rec => rec.getReadPairedFlag && rec.getSecondOfPairFlag }
+    (fragments, firstOfPair, secondOfPair)
+  }
+
+  /** Returns the next read name with format "<prefix>:<idx>", where "<prefix>" is either the supplied prefix or the
+    * longest common prefix of all read names, and "<idx>" is the 1-based consensus read index.  If no prefix was found,
+    * "CONSENSUS" is used.  If no records are given, the empty string is returned.
+    */
+  private def nextReadName(records: Seq[SAMRecord]): String = {
+    if (records.isEmpty) return ""
+    val curIdx = yieldAndThen(readIdx)(readIdx += 1)
+    val prefix = readNamePrefix.getOrElse(longestCommonPrefix(records.map(_.getReadName)).getOrElse("CONSENSUS"))
+    s"$prefix:$curIdx"
+  }
+
+  /** Attempts to create a consensus read from the given records adds it to `nextConsensusRecords`. Returns true if a
+    * consensus read was created, false otherwise. */
+  private def callConsensus(records: Seq[SAMRecord],
```

Comment:
`callConsensus` doesn't really imply side-effecting like adding it to the output 😬